connection-service/connection-flask.py

```diff
@@ -7,204 +7,244 @@
+  #  Store multiple connection ids and corresponding uris in a
+  #  dictionary associated with the appropriate partition.
+  timestamp=datetime.now()
+  js=json.loads(request.data)
+  if debug_level>2:
+    print (f"[{timestamp}] Publish {js=}")
+  part=js['partition']
+
+  if debug_level>1:
+    print(f"[{timestamp}] Publish {len(js['connections'])} connections in partition {part}"
+          f" from {request.remote_addr} uri={js['connections'][0]['uri']} ...")
+  partlock.acquire()
   if part in partitions:
     store=partitions[part]
   else:
-    store={'endpoint': {}, 'connection':{}, 'connection_type':{}}
-
-  if 'endpoint' in request.form:
-    ep=json.loads(request.form['endpoint'])
-    ep=json.dumps(ep)
-    store['endpoint'][ep]=request.form['uri']
-    if 'connection_type' in request.form:
-      store['connection_type'][ep]=request.form['connection_type']
-  if 'connection' in request.form:
-    conn=json.loads(request.form['connection'])
-    ep=json.dumps(conn['bind_endpoint'])
-    store['endpoint'][ep]=conn['uri']
-    store['connection'][ep]=conn
-
-  partitions[part]=store
-  cmgr.update(partitions)
+    store={}
+    partitions[part]=store
+    global maxpartitions
+    if len(partitions)>maxpartitions:
+      maxpartitions=len(partitions)
+    if not part in maxentries:
+      #print(f"Setting maxentries[{part}] to 0")
+      maxentries[part]=0
+
+  Connection=namedtuple(
+    'Connection',['uri','data_type','connection_type','time'])
+
+  for connection in js['connections']:
+    #print (f"{connection=}")
+    if 'uid' in  connection and 'uri' in connection:
+      uid=connection['uid']
+      store[uid]=Connection(uri=connection['uri'],
+                            connection_type=connection['connection_type'],
+                            data_type=connection['data_type'],
+                            time=timestamp)
+  now=datetime.now()
+  elapsed=now-timestamp
+  if debug_level>0:
+    print(f"[{now}] Publish took {elapsed.microseconds} us to add {len(js['connections'])} connections")
+  global npublishes, publish_time
+  publish_time+=elapsed
+  npublishes+=1
+  if len(store)>maxentries[part]:
+    maxentries[part]=len(store)
+
+  partlock.release()
   return 'OK'
 
 @app.route("/retract-partition",methods=['POST'])
 def retract_partition():
-  #print(f"retract_partition() request=[{request.form}]")
+  if debug_level>2:
+    print(f"[datetime.now()] retract_partition() request=[{request.form}]")
+  if 'partition' not in request.form:
+    abort(400)
   part=request.form['partition']
+  partlock.acquire()
   if part in partitions:
     partitions.pop(part)
-    cmgr.delete(part)
+    partlock.release()
     return 'OK'
   else:
+    partlock.release()
     abort(404)
 
 @app.route("/retract",methods=['POST'])
 def retract():
-  print(f"retract() request=[{request.form}]")
-  part=request.form['partition']
+  js=json.loads(request.data)
+  good=True
+  part=js['partition']
+  partlock.acquire()
```

Comment:
Could we have the happy path not nested too much here?
Something like this:

```
if part not in partitions:
   return make_response(f"partition {part} not found", 404)
...
```

Reply:
OK, not forgetting to release the lock of course.

Reply:
Looks like we could use a context manager:
```Py
  with partlock:
    # do stuff while locked
```